flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java

```diff
@@ -0,0 +1,342 @@
+	public final void invoke(IN value) throws Exception {
+		invoke(currentTransaction, value);
+	}
+
+	@Override
+	public final void notifyCheckpointComplete(long checkpointId) throws Exception {
+		// the following scenarios are possible here
+		//
+		//  (1) there is exactly one transaction from the latest checkpoint that
+		//      was triggered and completed. That should be the common case.
+		//      Simply commit that transaction in that case.
+		//
+		//  (2) there are multiple pending transactions because one previous
+		//      checkpoint was skipped. That is a rare case, but can happen
+		//      for example when:
+		//
+		//        - the master cannot persist the metadata of the last
+		//          checkpoint (temporary outage in the storage system) but
+		//          could persist a successive checkpoint (the one notified here)
+		//
+		//        - other tasks could not persist their status during
+		//          the previous checkpoint, but did not trigger a failure because they
+		//          could hold onto their state and could successfully persist it in
+		//          a successive checkpoint (the one notified here)
+		//
+		//      In both cases, the prior checkpoint never reach a committed state, but
+		//      this checkpoint is always expected to subsume the prior one and cover all
+		//      changes since the last successful one As a consequence, we need to commit
+		//      all pending transactions.
+		//
+		//  (3) Multiple transactions are pending, but the checkpoint complete notification
+		//      relates not to the latest. That is possible, because notification messages
+		//      can be delayed (in an extreme case till arrive after a succeeding checkpoint
+		//      was triggered) and because there can be concurrent overlapping checkpoints
+		//      (a new one is started before the previous fully finished).
+		//
+		// ==> There should never be a case where we have no pending transaction here
+		//
+
+		Iterator<TransactionAndCheckpoint<TXN>> pendingTransactionsIterator = pendingCommitTransactions.iterator();
+		checkState(pendingTransactionsIterator.hasNext(), "checkpoint completed, but no transaction pending");
+
+		List<TransactionAndCheckpoint<TXN>> remainingTransactions = new ArrayList<>();
+
+		for (TransactionAndCheckpoint<TXN> pendingTransaction : pendingCommitTransactions) {
+			if (pendingTransaction.checkpointId > checkpointId) {
+				remainingTransactions.add(pendingTransaction);
+				continue;
+			}
+
+			LOG.info("{} - checkpoint {} complete, committing completed checkpoint transaction {}",
+				name(), checkpointId, pendingTransaction);
+
+			// If this fails, there is actually a data loss
+			commit(pendingTransaction.transaction);
+
+			LOG.debug("{} - committed checkpoint transaction {}", name(), pendingTransaction);
+		}
+
+		pendingCommitTransactions.clear();
+		for (TransactionAndCheckpoint<TXN> remainingTransaction : remainingTransactions) {
+			pendingCommitTransactions.add(remainingTransaction);
```

Comment:
this seems redundant. It should be able to simply be removal while iterating, no?